daemon/graphdriver/devmapper/deviceset.go

```diff
@@ -1744,10 +1752,99 @@ func (devices *DeviceSet) AddDevice(hash, baseHash string) error {
 		return fmt.Errorf("devmapper: device %s already exists. Deleted=%v", hash, info.Deleted)
 	}
 
-	if err := devices.createRegisterSnapDevice(hash, baseInfo); err != nil {
+	devinfo := &devInfo{}
+
+	if err := devices.parseStorageOpt(storageOpt, devinfo); err != nil {
+		return err
+	}
+
+	if err := devices.createRegisterSnapDevice(hash, baseInfo, devinfo.Size); err != nil {
 		return err
 	}
 
+	// Grow the container rootfs.
+	if devinfo.Size > 0 {
+		if err := devices.growFS(hash); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (devices *DeviceSet) parseStorageOpt(storageOpt []string, devinfo *devInfo) error {
+
+	// Read size to change the block device size per container.
+	var size int64
+	var err error
+	for _, option := range storageOpt {
+		val := strings.SplitN(option, ":", 2)
+		key := strings.ToLower(val[0])
+		switch key {
+		case "size":
+			size, err = units.RAMInBytes(val[1])
+			if err != nil {
+				return err
+			}
+			devinfo.Size = uint64(size)
+		default:
+			return fmt.Errorf("Unknown option %s\n", key)
+		}
+	}
+
+	return nil
+}
+
+func (devices *DeviceSet) growFS(hash string) error {
+
+	info, err := devices.lookupDevice(hash)
+	if info == nil {
+		return fmt.Errorf("Failed to lookup device %s:%v", hash, err)
+	}
+
+	info.lock.Lock()
+	defer info.lock.Unlock()
+
+	if err := devices.activateDeviceIfNeeded(info, false); err != nil {
+		return fmt.Errorf("Error activating devmapper device for '%s': %s", hash, err)
+	}
+
+	defer devices.deactivateDevice(info)
+
+	fsMountPoint := "/run/docker/mnt"
```

Comment:
I thought it is simpler. We don't have to create another directory inside /run. Also I am not sure if all the systems where docker runs have /run or not. And also not sure why there is a need for unprivileged users to be able to see this temporary mount point.